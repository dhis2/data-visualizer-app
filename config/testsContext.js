import { render } from '@testing-library/react'
import PropTypes from 'prop-types'
import React from 'react'
import { Provider } from 'react-redux'

export function getStubContext() {
    return {
        i18n: {
            t: () => {},
        },
        store: {
            dispatch: () => {},
        },
    }
}

export const renderWithProviders = (ui, store, renderOptions = {}) => {
    const Wrapper = ({ children }) => (
        <Provider store={store}>{children}</Provider>
    )

    Wrapper.propTypes = {
        children: PropTypes.node,
    }

    return {
        store,
        ...render(ui, { wrapper: Wrapper, ...renderOptions }),
    }
}
