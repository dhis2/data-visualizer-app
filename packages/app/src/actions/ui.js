import {
    SET_UI,
    CLEAR_UI,
    SET_UI_FROM_VISUALIZATION,
    SET_UI_TYPE,
    SET_UI_OPTIONS,
    SET_UI_LAYOUT,
    ADD_UI_LAYOUT_DIMENSIONS,
    REMOVE_UI_LAYOUT_DIMENSIONS,
    SET_UI_ITEMS,
    ADD_UI_ITEMS,
    REMOVE_UI_ITEMS,
    ADD_UI_PARENT_GRAPH_MAP,
    SET_UI_PARENT_GRAPH_MAP,
    SET_UI_ACTIVE_MODAL_DIALOG,
    SET_UI_YEAR_ON_YEAR_SERIES,
    SET_UI_YEAR_ON_YEAR_CATEGORY,
} from '../reducers/ui';

export const acSetUi = value => ({
    type: SET_UI,
    value,
});

export const acClear = () => ({
    type: CLEAR_UI,
});

export const acSetUiFromVisualization = value => ({
    type: SET_UI_FROM_VISUALIZATION,
    value,
});

export const acSetUiType = value => ({
    type: SET_UI_TYPE,
    value,
});

export const acSetUiOptions = value => ({
    type: SET_UI_OPTIONS,
    value,
});

export const acSetUiLayout = value => ({
    type: SET_UI_LAYOUT,
    value,
});

export const acAddUiLayoutDimensions = value => ({
    type: ADD_UI_LAYOUT_DIMENSIONS,
    value,
});

export const acRemoveUiLayoutDimensions = value => ({
    type: REMOVE_UI_LAYOUT_DIMENSIONS,
    value,
});

export const acSetUiItems = value => ({
    type: SET_UI_ITEMS,
    value,
});

export const acAddUiItems = value => ({
    type: ADD_UI_ITEMS,
    value,
});

export const acRemoveUiItems = value => ({
    type: REMOVE_UI_ITEMS,
    value,
});

export const acSetUiYearOnYearSeries = value => ({
    type: SET_UI_YEAR_ON_YEAR_SERIES,
    value,
});

export const acSetUiYearOnYearCategory = value => ({
    type: SET_UI_YEAR_ON_YEAR_CATEGORY,
    value,
});

export const acAddParentGraphMap = value => ({
    type: ADD_UI_PARENT_GRAPH_MAP,
    value,
});

export const acSetParentGraphMap = value => ({
    type: SET_UI_PARENT_GRAPH_MAP,
    value,
});

export const acSetUiActiveModalDialog = value => ({
    type: SET_UI_ACTIVE_MODAL_DIALOG,
    value,
});
