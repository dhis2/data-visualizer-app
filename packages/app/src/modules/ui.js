import { YEAR_OVER_YEAR_LINE, YEAR_OVER_YEAR_COLUMN } from './chartTypes';
import {
    AXIS_NAME_COLUMNS,
    AXIS_NAME_ROWS,
    AXIS_NAME_FILTERS,
    getDimensionIdsByAxis,
    getItemIdsByDimension,
} from './layout';
import { FIXED_DIMENSIONS } from './fixedDimensions';
import { isYearOverYear } from './chartTypes';
import { getOptionsFromVisualization } from './options';

const peId = FIXED_DIMENSIONS.pe.id;

// Transform from backend model to store.ui format
export const getUiFromVisualization = (vis, currentState = {}) => ({
    ...currentState,
    type: vis.type,
    options: getOptionsFromVisualization(vis),
    layout: getDimensionIdsByAxis(vis),
    itemsByDimension: getItemIdsByDimension(vis),
    parentGraphMap: vis.parentGraphMap || currentState.parentGraphMap,
    yearOverYearSeries:
        isYearOverYear(vis.type) && vis.yearlySeries
            ? vis.yearlySeries
            : currentState.yearOverYearSeries,
    yearOverYearCategory: isYearOverYear(vis.type)
        ? vis.rows[0].items.map(item => item.id)
        : currentState.yearOverYearCategory,
});

// Transform from store.ui to year on year format
const yearOverYearUiAdapter = ui => {
    const state = Object.assign({}, ui);

    const items = Object.assign({}, state.itemsByDimension);
    delete items[peId];

    return {
        ...state,
        layout: {
            [AXIS_NAME_COLUMNS]: [],
            [AXIS_NAME_ROWS]: [],
            [AXIS_NAME_FILTERS]: [
                ...state.layout[AXIS_NAME_FILTERS],
                ...state.layout[AXIS_NAME_COLUMNS],
                ...state.layout[AXIS_NAME_ROWS],
            ].filter(dim => dim !== peId),
        },
        itemsByDimension: items,
    };
};

export const getAdaptedUiByType = ui => {
    switch (ui.type) {
        case YEAR_OVER_YEAR_LINE:
        case YEAR_OVER_YEAR_COLUMN: {
            return yearOverYearUiAdapter(ui);
        }
        default:
            return ui;
    }
};
