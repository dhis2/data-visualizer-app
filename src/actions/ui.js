import { getDisabledOptions } from '../modules/disabledOptions.js'
import {
    SET_UI,
    CLEAR_UI,
    CLEAR_UI_DATA_SORTING,
    SET_UI_FROM_VISUALIZATION,
    SET_UI_TYPE,
    SET_UI_DISABLED_OPTIONS,
    SET_UI_OPTIONS,
    SET_UI_LAYOUT,
    ADD_UI_LAYOUT_DIMENSIONS,
    REMOVE_UI_LAYOUT_DIMENSIONS,
    SET_UI_ITEMS,
    REMOVE_UI_ITEMS,
    ADD_UI_PARENT_GRAPH_MAP,
    SET_UI_ACTIVE_MODAL_DIALOG,
    SET_UI_YEAR_ON_YEAR_SERIES,
    SET_UI_YEAR_ON_YEAR_CATEGORY,
    SET_UI_ITEM_ATTRIBUTES,
    REMOVE_UI_ITEM_ATTRIBUTES,
    TOGGLE_UI_RIGHT_SIDEBAR_OPEN,
    SET_UI_RIGHT_SIDEBAR_OPEN,
    CLEAR_SERIES_TYPE,
    UPDATE_UI_SERIES_ITEM,
    SET_UI_OPTION,
    SET_UI_OPTION_FONT_STYLE,
    SET_UI_DATA_SORTING,
    sGetUiType,
    sGetUiOptions,
} from '../reducers/ui.js'

export const acSetUi = (value) => ({
    type: SET_UI,
    value,
})

export const acClearUi = (value) => ({
    type: CLEAR_UI,
    value,
})

export const acSetUiFromVisualization = (value) => ({
    type: SET_UI_FROM_VISUALIZATION,
    value,
})

export const acSetUiType = (value) => ({
    type: SET_UI_TYPE,
    value,
})

export const acSetUiDisabledOptions = (value) => ({
    type: SET_UI_DISABLED_OPTIONS,
    value,
})

export const acSetUiOptions = (value) => ({
    type: SET_UI_OPTIONS,
    value,
})

export const acSetUiOption = (value) => ({
    type: SET_UI_OPTION,
    value,
})

export const acSetUiOptionFontStyle = (value) => ({
    type: SET_UI_OPTION_FONT_STYLE,
    value,
})

export const acClearUiDataSorting = () => ({
    type: CLEAR_UI_DATA_SORTING,
})

export const acSetUiDataSorting = (value) => ({
    type: SET_UI_DATA_SORTING,
    value,
})

export const acUpdateUiSeriesItem = (value) => ({
    type: UPDATE_UI_SERIES_ITEM,
    value,
})

export const acSetUiLayout = (value) => ({
    type: SET_UI_LAYOUT,
    value,
})

export const acAddUiLayoutDimensions = (value) => ({
    type: ADD_UI_LAYOUT_DIMENSIONS,
    value,
})

export const acRemoveUiLayoutDimensions = (value) => ({
    type: REMOVE_UI_LAYOUT_DIMENSIONS,
    value,
})

export const acSetUiItems = (value) => ({
    type: SET_UI_ITEMS,
    value,
})

export const acRemoveUiItems = (value) => ({
    type: REMOVE_UI_ITEMS,
    value,
})

export const acSetUiYearOverYearSeries = (value) => ({
    type: SET_UI_YEAR_ON_YEAR_SERIES,
    value,
})

export const acSetUiYearOverYearCategory = (value) => ({
    type: SET_UI_YEAR_ON_YEAR_CATEGORY,
    value,
})

export const acSetUiItemAttributes = (value) => ({
    type: SET_UI_ITEM_ATTRIBUTES,
    value,
})

export const acRemoveUiItemAttributes = (value) => ({
    type: REMOVE_UI_ITEM_ATTRIBUTES,
    value,
})

export const acAddParentGraphMap = (value) => ({
    type: ADD_UI_PARENT_GRAPH_MAP,
    value,
})

export const acSetUiActiveModalDialog = (value) => ({
    type: SET_UI_ACTIVE_MODAL_DIALOG,
    value,
})

export const acToggleUiRightSidebarOpen = () => ({
    type: TOGGLE_UI_RIGHT_SIDEBAR_OPEN,
})

export const acSetUiRightSidebarOpen = () => ({
    type: SET_UI_RIGHT_SIDEBAR_OPEN,
})

export const acClearSeriesType = () => ({
    type: CLEAR_SERIES_TYPE,
})

export const tSetUiOptionAndDisabledOptions =
    (option) => (dispatch, getState) => {
        dispatch(acSetUiOption(option))

        const visType = sGetUiType(getState())
        const options = sGetUiOptions(getState())

        dispatch(
            acSetUiDisabledOptions(getDisabledOptions({ visType, options }))
        )
    }
