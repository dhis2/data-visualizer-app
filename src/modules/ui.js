import {
    DIMENSION_ID_PERIOD,
    DIMENSION_ID_ORGUNIT,
    DIMENSION_ID_DATA,
    layoutGetAxisIdDimensionIdsObject,
    layoutGetDimensionIdItemIdsObject,
    VIS_TYPE_YEAR_OVER_YEAR_LINE,
    VIS_TYPE_YEAR_OVER_YEAR_COLUMN,
    VIS_TYPE_PIVOT_TABLE,
    VIS_TYPE_SCATTER,
    defaultVisType,
    isYearOverYear,
    getAdaptedUiLayoutByType,
    VIS_TYPE_GAUGE,
    VIS_TYPE_SINGLE_VALUE,
} from '@dhis2/analytics'
import { BASE_FIELD_YEARLY_SERIES } from './fields/baseFields.js'
import { getInverseLayout } from './layout.js'
import { getOptionsFromVisualization } from './options.js'
import { removeLastPathSegment } from './orgUnit.js'

export const SERIES_ITEM_TYPE_PROP = 'type'
export const SERIES_ITEM_AXIS_PROP = 'axis'
export const ITEM_ATTRIBUTE_VERTICAL = 'VERTICAL'
export const ITEM_ATTRIBUTE_HORIZONTAL = 'HORIZONTAL'

// Transform from backend model to store.ui format
export const getUiFromVisualization = (vis, currentState = {}) => ({
    ...currentState,
    type: vis.type || defaultVisType,
    options: getOptionsFromVisualization(vis),
    layout: layoutGetAxisIdDimensionIdsObject(vis),
    itemsByDimension: layoutGetDimensionIdItemIdsObject(vis),
    parentGraphMap:
        vis.parentGraphMap ||
        getParentGraphMapFromVisualization(vis) ||
        currentState.parentGraphMap,
    yearOverYearSeries:
        isYearOverYear(vis.type) && vis[BASE_FIELD_YEARLY_SERIES]
            ? vis[BASE_FIELD_YEARLY_SERIES]
            : currentState.yearOverYearSeries,
    yearOverYearCategory: isYearOverYear(vis.type)
        ? vis.rows[0].items.map((item) => item.id)
        : currentState.yearOverYearCategory,
})

// Transform from store.ui to default format
const defaultUiAdapter = (ui) => ({
    ...ui,
    layout: getAdaptedUiLayoutByType(ui.layout, ui.type),
})

// Transform from store.ui to year on year format
const yearOverYearUiAdapter = (ui) => {
    const state = Object.assign({}, ui)

    const items = Object.assign({}, state.itemsByDimension)
    delete items[DIMENSION_ID_PERIOD]

    return {
        ...state,
        layout: getAdaptedUiLayoutByType(ui.layout, ui.type),
        itemsByDimension: items,
    }
}

const singleValueUiAdapter = (ui) => {
    const adaptedUi = defaultUiAdapter(ui)
    adaptedUi.options.measureCriteria = undefined
    return adaptedUi
}

// Transform from store.ui to scatter format
const scatterUiAdapter = (ui) => {
    const adaptedUi = {
        ...ui,
        layout: getAdaptedUiLayoutByType(ui.layout, ui.type),
    }

    const dataItems = ui.itemsByDimension[DIMENSION_ID_DATA] || []

    adaptedUi.itemAttributes = [
        // TODO: Should this be cleared for all other uiAdapters?
        ...(dataItems[0]
            ? [{ id: dataItems[0], attribute: ITEM_ATTRIBUTE_VERTICAL }]
            : []),
        ...(dataItems[1]
            ? [{ id: dataItems[1], attribute: ITEM_ATTRIBUTE_HORIZONTAL }]
            : []),
    ]

    const items = Object.assign({}, ui.itemsByDimension)
    items[DIMENSION_ID_DATA] = dataItems.slice(0, 2)

    adaptedUi.itemsByDimension = items

    adaptedUi.options.measureCriteria = undefined

    return adaptedUi
}

export const getAdaptedUiByType = (ui) => {
    switch (ui.type) {
        case VIS_TYPE_YEAR_OVER_YEAR_LINE:
        case VIS_TYPE_YEAR_OVER_YEAR_COLUMN: {
            return yearOverYearUiAdapter(ui)
        }
        case VIS_TYPE_PIVOT_TABLE:
            return ui
        case VIS_TYPE_GAUGE:
        case VIS_TYPE_SINGLE_VALUE:
            return singleValueUiAdapter(ui)
        case VIS_TYPE_SCATTER:
            return scatterUiAdapter(ui)
        default:
            return defaultUiAdapter(ui)
    }
}

export const getParentGraphMapFromVisualization = (vis) => {
    const dimensionIdsByAxis = layoutGetAxisIdDimensionIdsObject(vis)
    const inverseLayout = getInverseLayout(dimensionIdsByAxis)
    const ouAxis = inverseLayout[DIMENSION_ID_ORGUNIT]

    if (!ouAxis) {
        return {}
    }

    const parentGraphMap = {}
    const ouDimension = vis[ouAxis].find(
        (dimension) => dimension.dimension === DIMENSION_ID_ORGUNIT
    )

    ouDimension.items
        .filter((orgUnit) => orgUnit.path)
        .forEach((orgUnit) => {
            if ('/' + orgUnit.id === orgUnit.path) {
                // root org unit case
                parentGraphMap[orgUnit.id] = ''
            } else {
                const path = removeLastPathSegment(orgUnit.path)
                parentGraphMap[orgUnit.id] =
                    path[0] === '/' ? path.substr(1) : path
            }
        })

    return parentGraphMap
}

export const mergeUiMaps = (destinationMap, sourceMap, propName) => {
    Object.keys(sourceMap || {}).forEach((key) => {
        if (!(key in destinationMap)) {
            destinationMap[key] = {}
        }

        destinationMap[key][propName] = sourceMap[key]
    })
}
