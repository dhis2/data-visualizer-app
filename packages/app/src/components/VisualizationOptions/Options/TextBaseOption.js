import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TextField from '@material-ui/core/TextField';
import { sGetUiOptions } from '../../../reducers/ui';
import { acSetUiOptions } from '../../../actions/ui';

export const TextBaseOption = ({
    className,
    type,
    option,
    value,
    onChange,
}) => (
    <TextField
        className={className}
        type={type}
        label={option.label}
        onChange={event => onChange(event.target.value)}
        value={value}
        helperText={option.helperText}
    />
);

TextBaseOption.propTypes = {
    className: PropTypes.string,
    type: PropTypes.string,
    option: PropTypes.object,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    onChange: PropTypes.func,
};

const mapStateToProps = (state, ownProps) => ({
    value: sGetUiOptions(state)[ownProps.option.name] || '',
});

const mapDispatchToProps = (dispatch, ownProps) => ({
    onChange: value =>
        dispatch(acSetUiOptions({ [ownProps.option.name]: value })),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(TextBaseOption);
