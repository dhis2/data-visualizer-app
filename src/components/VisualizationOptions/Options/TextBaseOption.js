import { Checkbox, Box, Input, InputField } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { acSetUiOption } from '../../../actions/ui.js'
import { sGetUiOption, sGetUiDisabledOption } from '../../../reducers/ui.js'
import {
    tabSectionOption,
    tabSectionOptionToggleable,
} from '../styles/VisualizationOptions.style.js'
import TextStyle from './TextStyle.js'

export const UnconnectedTextBaseOption = ({
    type,
    label,
    placeholder,
    helpText,
    width,
    option,
    value,
    onChange,
    onToggle,
    fontStyleKey,
    toggleable,
    checked,
    inline,
    disabled,
    dataTest,
    isVertical,
}) => (
    <div className={inline ? '' : tabSectionOption.className}>
        {toggleable ? (
            <Checkbox
                checked={checked}
                label={label}
                name={`${option.name}-toggle`}
                onChange={({ checked }) => onToggle(checked)}
                dense
                disabled={disabled}
                dataTest={`${dataTest}-checkbox`}
            />
        ) : null}
        {!toggleable || checked ? (
            <div
                className={
                    toggleable ? tabSectionOptionToggleable.className : ''
                }
            >
                {inline ? (
                    <Box width={width}>
                        <Input
                            type={type}
                            onChange={({ value }) => onChange(value)}
                            name={option.name}
                            value={value.toString()}
                            placeholder={placeholder}
                            dense
                            disabled={disabled}
                            dataTest={`${dataTest}-input`}
                        />
                    </Box>
                ) : (
                    <InputField
                        type={type}
                        label={toggleable ? '' : label}
                        onChange={({ value }) => onChange(value)}
                        name={option.name}
                        value={value}
                        helpText={disabled ? '' : helpText}
                        placeholder={placeholder}
                        inputWidth={width}
                        dense
                        disabled={disabled}
                        dataTest={`${dataTest}-input`}
                    />
                )}
                {fontStyleKey ? (
                    <TextStyle
                        fontStyleKey={fontStyleKey}
                        disabled={disabled}
                        dataTest={`${dataTest}-text-style`}
                        axisId={option.axisId}
                        isVertical={isVertical}
                    />
                ) : null}
            </div>
        ) : null}
    </div>
)

UnconnectedTextBaseOption.defaultProps = {
    option: {},
}

UnconnectedTextBaseOption.propTypes = {
    checked: PropTypes.bool,
    dataTest: PropTypes.string,
    disabled: PropTypes.bool,
    fontStyleKey: PropTypes.string,
    helpText: PropTypes.string,
    inline: PropTypes.bool,
    isVertical: PropTypes.bool,
    label: PropTypes.string,
    option: PropTypes.object,
    placeholder: PropTypes.string,
    toggleable: PropTypes.bool,
    type: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    width: PropTypes.string,
    onChange: PropTypes.func,
    onToggle: PropTypes.func,
}

const mapStateToProps = (state, ownProps) => ({
    disabled: Boolean(sGetUiDisabledOption(state, ownProps.option)),
    helpText:
        sGetUiDisabledOption(state, ownProps.option)?.helpText ||
        ownProps.helpText,
    value: sGetUiOption(state, ownProps.option) || '',
    checked:
        sGetUiOption(state, {
            id: ownProps.option.enabledId,
            axisId: ownProps.option.axisId,
        }) || false,
})

const mapDispatchToProps = (dispatch, ownProps) => ({
    onChange: (value) =>
        dispatch(
            acSetUiOption({
                optionId: ownProps.option.id || ownProps.option.name,
                axisId: ownProps.option.axisId,
                value,
            })
        ),
    onToggle: (checked) =>
        dispatch(
            acSetUiOption({
                optionId: ownProps.option.enabledId,
                axisId: ownProps.option.axisId,
                value: checked,
            })
        ),
})

export const TextBaseOption = connect(
    mapStateToProps,
    mapDispatchToProps
)(UnconnectedTextBaseOption)
