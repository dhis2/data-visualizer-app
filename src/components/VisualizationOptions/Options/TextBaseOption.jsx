import { Checkbox, Box, Input, InputField } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { acSetUiOption } from '../../../actions/ui.js'
import { sGetUiOption, sGetUiDisabledOption } from '../../../reducers/ui.js'
import styles from '../styles/VisualizationOptions.module.css'
import TextStyle from './TextStyle.jsx'

const OPTION_PROP_DEFAULT = {}

const UnconnectedTextBaseOption = ({
    type,
    label,
    placeholder,
    helpText,
    width,
    option = OPTION_PROP_DEFAULT,
    value,
    onChange,
    onToggle,
    fontStyleKey,
    toggleable,
    checked,
    inline,
    disabled,
    dataTest,
    isVertical,
}) => (
    <div className={inline ? '' : styles.tabSectionOption}>
        {toggleable ? (
            <Checkbox
                checked={checked}
                label={label}
                name={`${option.name}-toggle`}
                onChange={({ checked }) => onToggle(checked)}
                dense
                disabled={disabled}
                dataTest={`${dataTest}-checkbox`}
            />
        ) : null}
        {!toggleable || checked ? (
            <div
                className={toggleable ? styles.tabSectionOptionToggleable : ''}
            >
                {inline ? (
                    <Box width={width}>
                        <Input
                            type={type}
                            onChange={({ value }) => onChange(value)}
                            name={option.name}
                            value={value.toString()}
                            placeholder={placeholder}
                            dense
                            disabled={disabled}
                            dataTest={`${dataTest}-input`}
                        />
                    </Box>
                ) : (
                    <InputField
                        type={type}
                        label={toggleable ? '' : label}
                        onChange={({ value }) => onChange(value)}
                        name={option.name}
                        value={value}
                        helpText={disabled ? '' : helpText}
                        placeholder={placeholder}
                        inputWidth={width}
                        dense
                        disabled={disabled}
                        dataTest={`${dataTest}-input`}
                    />
                )}
                {fontStyleKey ? (
                    <TextStyle
                        fontStyleKey={fontStyleKey}
                        disabled={disabled}
                        dataTest={`${dataTest}-text-style`}
                        axisId={option.axisId}
                        isVertical={isVertical}
                    />
                ) : null}
            </div>
        ) : null}
    </div>
)

UnconnectedTextBaseOption.propTypes = {
    checked: PropTypes.bool,
    dataTest: PropTypes.string,
    disabled: PropTypes.bool,
    fontStyleKey: PropTypes.string,
    helpText: PropTypes.string,
    inline: PropTypes.bool,
    isVertical: PropTypes.bool,
    label: PropTypes.string,
    option: PropTypes.object,
    placeholder: PropTypes.string,
    toggleable: PropTypes.bool,
    type: PropTypes.string,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    width: PropTypes.string,
    onChange: PropTypes.func,
    onToggle: PropTypes.func,
}

const mapStateToProps = (state, ownProps) => ({
    disabled: Boolean(sGetUiDisabledOption(state, ownProps.option)),
    helpText:
        sGetUiDisabledOption(state, ownProps.option)?.helpText ||
        ownProps.helpText,
    value: sGetUiOption(state, ownProps.option) || '',
    checked:
        sGetUiOption(state, {
            id: ownProps.option.enabledId,
            axisId: ownProps.option.axisId,
        }) || false,
})

const mapDispatchToProps = (dispatch, ownProps) => ({
    onChange: (value) =>
        dispatch(
            acSetUiOption({
                optionId: ownProps.option.id || ownProps.option.name,
                axisId: ownProps.option.axisId,
                value,
            })
        ),
    onToggle: (checked) =>
        dispatch(
            acSetUiOption({
                optionId: ownProps.option.enabledId,
                axisId: ownProps.option.axisId,
                value: checked,
            })
        ),
})

export const TextBaseOption = connect(
    mapStateToProps,
    mapDispatchToProps
)(UnconnectedTextBaseOption)
