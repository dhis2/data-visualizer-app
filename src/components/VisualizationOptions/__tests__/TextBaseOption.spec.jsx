import React from 'react'
import { render, screen, queryByAttribute } from '@testing-library/react'
import { TextBaseOption } from '../Options/TextBaseOption.jsx'
import { setupTestStore } from '../../../configureStore.js'
import { renderWithProviders } from '../../../../config/testsContext.js'
import { DEFAULT_UI } from '../../../reducers/ui.js'
import {
    OPTION_FONT_SIZE,
    OPTION_TARGET_LINE_ENABLED,
} from '../../../modules/options.js'

jest.mock('@dhis2/ui', () => ({
    ...jest.requireActual('@dhis2/ui'),
    Input: ({ name, value }) => <div id="input" name={name} value={value} />,
    InputField: ({ name, value }) => (
        <div id="inputfield" name={name} value={value} />
    ),
    Checkbox: ({ name, value }) => (
        <div id="checkbox" name={name} value={value} />
    ),
    Box: ({ children }) => <div id="box">{children}</div>,
}))

test('TextBaseOption renders an <InputField />', () => {
    const option = { name: 'title' }
    const initialState = {
        ui: {
            ...DEFAULT_UI,
            options: {
                ...DEFAULT_UI.options,
                title: 'hello',
            },
        },
    }
    const store = setupTestStore(initialState)

    const { container } = renderWithProviders(
        <TextBaseOption option={option} />,
        store
    )

    expect(container).toMatchSnapshot()
})

test('TextBaseOption renders a <Input /> field when inline is true', () => {
    const option = { name: 'title' }
    const initialState = {
        ui: {
            ...DEFAULT_UI,
            options: {
                ...DEFAULT_UI.options,
                title: 'hello',
            },
        },
    }
    const store = setupTestStore(initialState)

    const { container } = renderWithProviders(
        <TextBaseOption option={option} inline={true} toggleable={false} />,
        store
    )

    expect(container).toMatchSnapshot()
})

test('TextBaseOption renders a <Checkbox /> when toggleable is set to true', () => {
    const option = { name: 'title' }
    const store = setupTestStore({})

    const { container } = renderWithProviders(
        <TextBaseOption label="The label" option={option} toggleable={true} />,
        store
    )

    expect(container).toMatchSnapshot()
})

test('TextBaseOption renders a <InputField /> when the checkbox is checked', () => {
    const option = { enabledId: 'targetLineEnabled', axisId: 'x_0' }
    const initialState = {
        ui: {
            ...DEFAULT_UI,
            options: {
                ...DEFAULT_UI.options,
                axes: [
                    {
                        index: 0,
                        type: 'x',
                        targetLine: { enabled: true },
                    },
                ],
            },
        },
    }
    const store = setupTestStore(initialState)

    const { container } = renderWithProviders(
        <TextBaseOption label="The label" option={option} toggleable={false} />,
        store
    )

    expect(container).toMatchSnapshot()
})

test('TextBaseOption renders a TextStyle component when fontStyleKey is passed', () => {
    const option = { id: OPTION_FONT_SIZE, name: OPTION_FONT_SIZE }
    const helpText = 'Font size help'
    const initialState = {
        ui: {
            ...DEFAULT_UI,
            options: {
                ...DEFAULT_UI.options,
                [OPTION_FONT_SIZE]: 'LARGE',
            },
        },
    }
    const store = setupTestStore(initialState)

    const getByDataTest = queryByAttribute.bind(null, 'data-test')

    const { container } = renderWithProviders(
        <TextBaseOption
            option={option}
            helpText={helpText}
            fontStyleKey={OPTION_FONT_SIZE}
            dataTest="font-size-input"
        />,
        store
    )

    expect(container).toMatchSnapshot()

    const containerDiv = getByDataTest(container, 'font-size-input-text-style')
    expect(containerDiv).toBeInTheDocument()
})
