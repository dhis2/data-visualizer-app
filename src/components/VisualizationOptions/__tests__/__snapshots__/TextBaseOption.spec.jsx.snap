// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBaseOption renders a <Checkbox /> when toggleable is set to true 1`] = `
<div>
  <div>
    <div
      id="checkbox"
      name="title-toggle"
    />
  </div>
</div>
`;

exports[`TextBaseOption renders a <Input /> field when inline is true 1`] = `
<div>
  <div
    class=""
  >
    <div
      class=""
    >
      <div
        id="box"
      >
        <div
          id="input"
          name="title"
          value="hello"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TextBaseOption renders a <InputField /> when the checkbox is checked 1`] = `
<div>
  <div>
    <div
      class=""
    >
      <div
        id="inputfield"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextBaseOption renders a TextStyle component when fontStyleKey is passed 1`] = `
<div>
  <div>
    <div
      class=""
    >
      <div
        id="inputfield"
        name="fontSize"
        value="LARGE"
      />
      <div
        data-test="font-size-input-text-style"
      >
        <div
          class=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TextBaseOption renders an <InputField /> 1`] = `
<div>
  <div>
    <div
      class=""
    >
      <div
        id="inputfield"
        name="title"
        value="hello"
      />
    </div>
  </div>
</div>
`;
